rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1272 @@
+
+    # This is okay, indentation increases only one level per line
+    let
+      x = {
+        a = foo
+          bar
+          baz;
+      }; # <-- The decrease by two levels here is okay
+    in
+    null
+    ```
+
+### Expansion of expressions
+
+Unless stated otherwise, any expression that fits onto one single line must be trivially formatted as such.
+
+For list elements, attributes, and function arguments, the following applies:
+
+- If expanded into multiple lines, each item must be on its own line.
+  - Grouping similar items together can be done by adding blank lines or comments between the groups instead.
+  - This also applies to the first item, so e.g. `[ firstElement` in a multi line list is not allowed.
+- Long sequences of items should be liberally expanded, even if they would fit onto one line character-wise.
+  - The motivation is to keep the information per line manageable. Usually "number of elements" is a better metric for that than "line length".
+  - The cutoff is usually determined empirically based on common usage patterns.
+
+**Examples:**
+
+```nix
+{
+  buildInputs = [
+    foo
+    bar
+    baz
+
+    somethingElse
+  ];
+
+  systemd.services = {
+    foo = { };
+    bar = { };
+  };
+
+  inherit
+    lib
+    foo
+    bar
+    baz
+    ;
+}
+```
+
+### Function application
+
+- In a function application chain, the first element is treated as the "function" and the remaining ones as "arguments".
+- As many arguments as possible must be fit onto the first line.
+  - If all but the last argument do fit, then the last argument may also start on the first line.
+  - If an earlier argument does not fit onto the first line, then that argument and all the following ones must start on their own line.
+  - All arguments that are not on the same line as the function must be indented by one level.
+
+**Examples:**
+
+```nix
+# All arguments fit onto the first line
+function arg1 arg2
+
+# The line length limit is reached, so the remaining arguments need to be on their own lines
+function arg1 arg2 arg3
+  arg4
+  arg5
+
+# The last argument is a multiline expression, so it doesn't fit on the first line,
+# but it can still start on the first line
+function arg1 arg2 {
+  more = "things";
+}
+
+# The second argument doesn't fit on the first line, but it's not the last argument,
+# so it needs to start on a new line
+function arg1 arg2
+  {
+    more = "things";
+  }
+  arg3
+
+# Same with more multiline arguments
+function
+  {
+    a = 1;
+    b = 2;
+  }
+  {
+    c = 1;
+    d = 2;
+  }
+```
+
+**Drawbacks**
+
+- This style sometimes forces lists or attribute sets to start on a new line, with additional indentation of their items.
+
+**Alternatives**
+
+- Compacting multiline arguments like this:
+  ```nix
+  function arg1 {
+    # stuff
+  } arg3
+
+  function {
+    # …
+  } {
+    # …
+  }
+  ```
+  - This violates the guideline of the indentation representing the expression structure, and thus reduces readability.
```

Comment:
I think it would be nice if the first form was allowed.

I think being strict about this produces an unfortunate amount of indentation for no real readability gain.

```nix
    pkgsStatic.callPackage
      (
        {
          execline,
          kmod,
          mdevd,
          cryptsetup,
          util-linuxMinimal,
        }:

        # ...
      )
      { }
```

Could instead be

```nix
    pkgsStatic.callPackage (
      {
        execline,
        kmod,
        mdevd,
        cryptsetup,
        util-linuxMinimal,
      }:

      # ...
    ) { }
```

I think the structure of the expression is still perfectly clear and represented by the indentation in this case.

Reply:
One practical way of achieving this could be to allow a _single_ argument to be expanded over multiple lines, but other rules to achieve the same result would also work.

Reply:
What you propose would be acceptable to me *as long as the last argument is single-line*. In fact, we already have a similar special case for bindings with `//` operators.

Current workarounds to this which is already used in nixpkgs is to use `flip` to put the largest argument last (although I'm personally not that fond of it). Once this RFC is merged I hope continue working on the pipe operator RFC, because many of the cases with ugly formatting could be written in a cleaner way using pipes.

Reply:
> What you propose would be acceptable to me _as long as the last argument is single-line_. In fact, we already have a similar special case for bindings with `//` operators.

That would be perfectly fine for me.

Reply:
I've pushed a first draft implementation to the nixfmt branch, feel free to try it out. I think the simplification you proposed is an improvement in many cases, however I still need to tweak the rules to sort out some false positives where it shouldn't be applied.

Reply:
Oh, sorry, I misread your proposal.  "as long as the last argument is single-line" doesn't work for callPackage, because you usually end with a `{ }` after your big package definition function.  So this doesn't actually solve the example I proposed.

Reply:
(The file I'm formatting here is https://spectrum-os.org/git/spectrum/tree/host/initramfs/default.nix?id=1a58ed4c6b9b5fbfb3f8d2aad36cf84dfb77ac58 if it's helpful to reproduce)

Reply:
I ditched the prototype implementation and did lots of refactoring to implement this properly. It should also handle your examples correctly now. Nixpkgs diff: https://github.com/piegamesde/nixpkgs/pull/4/commits/7a7a7b5036840cff8e2f7f09032180ffa1d83a79

Reply:
This is _so_ much tighter without losing any readability. Very nice!

Reply:
I concour! +1 to this change! =)

Reply:
I really liked the way function applications with complex arguments were formatted before. It created a clear visual separation between the arguments.

Reply:
I agree that while the change is mostly positive, it sometimes reduces readability. I will try to find some heuristic for cases where this should better not be applied. So if you see any changes you perceive as regressions, feel free to point them out here for sample size.